library-best-practices/JLBP-3.md

```diff
@@ -0,0 +1,45 @@
+[JLBP-3] Use Semantic versioning
+--------------------------------
+
+Definition of semver: https://semver.org
+
+- For the purpose of compliance with semver, the stable surface of a library
+  serves as its "Public API." Annotations can be used to mark features
+  (classes, methods, etc) as unstable, and thus not part of the Public API,
+  such that semver rules don't have to apply. 
+  - Examples of annotations:
+    - Guava uses `@Beta`
+    - grpc-java uses `@ExperimentalApi`
+    - Google-cloud-java uses `@BetaApi`, `@InternalApi`, and
+      `@InternalExtensionOnly`
+  - Library documentation should let users know how to detect usage of
```

Comment:
"let users know how to detect usage of" --> clearly mark and warn users of

Reply:
The "detect" part of this sentence references usage of tools like https://github.com/google/guava-beta-checker - does that change your recommendation on how to word this?

Reply:
I would expect library documentation to specify the incompatible changes, not suggest tools for users to find out what the developer already knows.

Reply:
But... the idea is to do both (documentation + tooling), they are not mutually exclusive. We certainly do need the documentation - that is not in question. The tool part is the same idea as `@Deprecated`, where users can choose to have warnings or build failures if they use deprecated features; we also want to enable users to get warnings or build failures if they use unstable features (it's their choice). It's easy for developers to accidentally use things because of IDE autocomplete, and not realize what they have done. Also, it's something that junior developers on a team can easily do without realizing the consequences, and it's much easier on the senior developers to enforce project rules using tools instead of having to very carefully inspect all changes being made.

Reply:
As written this seems to mix up the recommendation to library vendors and library users. Or perhaps it's a meta-recommendation: we're recommending that library vendors should recommend these tools to their users.

Reply:
Yes, it's a meta-recommendation, that's a good way to describe it. Recommendations for library users will be documented separately from the JLBP series. I will try to update the text to make this clear.